google-cloud-firestore/src/main/java/com/google/cloud/firestore/UpdateBuilder.java

```diff
@@ -723,40 +735,88 @@ BatchState getState() {
     return state;
   }
 
-  Set<DocumentReference> getDocuments() {
-    return documents;
+  boolean hasPath(String path) {
+    return this.pendingOperations.containsKey(path);
+  }
+
+  Set<String> getPendingDocs() {
+    return this.pendingOperations.keySet();
+  }
+
+  void markComplete() {
+    completeFuture.set(null);
   }
 
-  int getOperationCount() {
-    return resultsMap.size();
+  int getPendingOperationCount() {
+    return pendingOperations.size();
+  }
+
+  /**
+   * Removes all operations not specified in documentsToRetry from the batch. Marks the batch as
+   * READY_TO_SEND in order to allow the next batch to be sent.
+   */
+  void sliceBatchForRetry(final Set<String> documentsToRetry) {
+    Iterator<WriteOperation> iterator = writes.iterator();
+    while (iterator.hasNext()) {
+      WriteOperation operation = iterator.next();
+      if (!documentsToRetry.contains(operation.path)) {
+        iterator.remove();
+      }
+    }
+
+    // Mark the batch as ready to send in order to allow the batch to be retried again.
+    state = BatchState.READY_TO_SEND;
   }
 
   private ApiFuture<WriteResult> processOperation(DocumentReference documentReference) {
     Preconditions.checkState(
-        allowDuplicateDocs() || !documents.contains(documentReference.getPath()),
+        allowDuplicateDocs() || !pendingOperations.containsKey(documentReference.getPath()),
         "Batch should not contain writes to the same document");
     Preconditions.checkState(state == BatchState.OPEN, "Batch should be OPEN when adding writes");
-    documents.add(documentReference);
     SettableApiFuture<WriteResult> result = SettableApiFuture.create();
-    resultsMap.put(getOperationCount(), result);
+    pendingOperations.put(documentReference.getPath(), result);
 
-    if (getOperationCount() == maxBatchSize) {
+    if (getPendingOperationCount() == maxBatchSize) {
       state = BatchState.READY_TO_SEND;
     }
 
     return result;
   }
   /** Resolves the individual operations in the batch with the results. */
   void processResults(List<BatchWriteResult> results, @Nullable Exception error) {
-    for (int i = 0; i < resultsMap.size(); ++i) {
-      if (error == null) {
-        SettableApiFuture<WriteResult> future = resultsMap.get(i);
-        convertBatchWriteResult(results.get(i), future);
-      } else {
-        resultsMap.get(i).setException(error);
+    for (BatchWriteResult result : results) {
+      if (result.getException() != null) {
+        pendingOperations.get(result.getKey()).setException(result.getException());
+        pendingOperations.remove(result.getKey());
+      } else if (result.getStatus() == Status.OK || !shouldRetry(result.getStatus())) {
```

Comment:
We should really clean up the BatchWriteStatus to only offer one way to specify success or exceptions. This code is a bit confusing as exceptions are handled in two different places based on either the exception or the status code.